### PR TITLE
Administration functionalities

### PR DESCRIPTION
**Pull request by morteza-araby**
1- Get Request to get list of users for administrator
2- Post request to reset password for users
3 - Get Request to get a paginated users With start and limit.
4 - Search users from database and return an array of found users or empty

### CODE REVIEW

**Comment by erikjohnston** [comment]
@matrixbot ok to test

**Comment by erikjohnston** [comment]
There's a bunch of PEP8 violations:

```
synapse/handlers/admin.py:22:1: F401 'operator.attrgetter' imported but unused
synapse/rest/client/v1/admin.py:40:9: F841 local variable 'auth_user' is assigned to but never used
synapse/rest/client/v1/admin.py:57:1: E302 expected 2 blank lines, found 1
synapse/rest/client/v1/admin.py:159:1: E302 expected 2 blank lines, found 1
synapse/rest/client/v1/admin.py:190:1: E302 expected 2 blank lines, found 1
synapse/rest/client/v1/admin.py:204:9: F841 local variable 'auth_user' is assigned to but never used
synapse/rest/client/v1/admin.py:253:1: E302 expected 2 blank lines, found 1
synapse/rest/client/v1/admin.py:290:1: E302 expected 2 blank lines, found 1
synapse/storage/_base.py:931:37: E128 continuation line under-indented for visual indent
synapse/storage/_base.py:951:91: E501 line too long (92 > 90 characters)
synapse/storage/_base.py:953:91: E501 line too long (93 > 90 characters)
synapse/storage/_base.py:973:91: E501 line too long (92 > 90 characters)
synapse/storage/_base.py:1039:1: E303 too many blank lines (3)
```

You should be able to test locally by installing flake8 and then running `flake8 synapse/`

File: scripts-dev/add_admin.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import sqlite3
+import logging
+
+
+def reg_admin(cursor):
+    headers = {'content-type': 'application/json'}
+    url = 'http://localhost:8008/_matrix/client/api/v1/register'
+    post_data = {'user': 'admin',
+                 'password': 'admin123',
+                 'type': 'm.login.password'}
```
Comment (erikjohnston) [inline scripts-dev/add_admin.py]: I'm a bit hesitant to add example scripts like this, as they can't directly be run. I believe the `register_new_user` script actually has an option to specify the new user is an admin or not.

**Comment by erikjohnston** [comment]
(The sytest dendron failure is probably spurious, ignore it)

**Comment by morteza-araby** [comment]
Hi,
Made a new commit.
1- Fix all error from flake8
2- add docstring to all the new methods
3 - Used _simple_select_list_paginate as private method which is more
generic to use for other list search as well and
used get_user_list_paginate for calling from rest api

File: scripts-dev/add_admin.py (hunk 1 above)
Comment (morteza-araby) [inline scripts-dev/add_admin.py]: Hi, 
Made a new commit.
1- Fix all error from flake8
2- add docstring to all the new methods
3 - Used _simple_select_list_paginate as private method which is more generic to use for other list search as well and used get_user_list_paginate for calling from rest api

**Comment by richvdh** [comment]
It'd be nice if someone could write something in https://github.com/matrix-org/synapse/tree/master/docs/admin_api about this.

**Comment by jcgruenhage** [comment]
Also, it would be great if the swagger ui supported this.

**Comment by richvdh** [comment]
The swagger spec is limited to the client-server API. These APIs are synapse-specific.

Of course, you *could* give them a swagger UI, but it wouldn't be appropriate to include them in the current one.

**Comment by jcgruenhage** [comment]
It would probably be more appropriate to write a little synapse admin client instead
(Or maybe even a web interface for those :D)

File: synapse/rest/client/v1/admin.py [hunk 2]
```diff
@@ -25,6 +26,34 @@
 logger = logging.getLogger(__name__)
 
 
+class UsersRestServlet(ClientV1RestServlet):
+    PATTERNS = client_path_patterns("/admin/users/(?P<user_id>[^/]*)")
```
Comment (richvdh) [inline synapse/rest/client/v1/admin.py]: is my understanding correct? this takes a user_id, and then ignores it? @morteza-araby

File: synapse/rest/client/v1/admin.py [hunk 3]
```diff
@@ -25,6 +26,34 @@
...
+
+        if not is_admin:
+            raise AuthError(403, "You are not a server admin")
+
+        # To allow all users to get the users list
+        # if not is_admin and target_user != auth_user:
+        #     raise AuthError(403, "You are not a server admin")
+
+        if not self.hs.is_mine(target_user):
+            raise SynapseError(400, "Can only users a local user")
```
Comment (richvdh) [inline synapse/rest/client/v1/admin.py]: this error makes no sense

File: synapse/rest/client/v1/admin.py [hunk 4]
```diff
@@ -128,8 +157,199 @@ def on_POST(self, request, target_user_id):
...
+
+        if not is_admin:
+            raise AuthError(403, "You are not a server admin")
+
+        params = parse_json_object_from_request(request)
+        new_password = params['new_password']
+        if not new_password:
+            raise SynapseError(400, "Missing 'new_password' arg")
+
+        logger.info("new_password: %r", new_password)
```
Comment (richvdh) [inline synapse/rest/client/v1/admin.py]: we *really* shouldn't log passwords at info.

File: synapse/rest/client/v1/admin.py (hunk 2 above)
Comment (morteza-araby) [inline synapse/rest/client/v1/admin.py]: you are right, I must have missed that. The UsersRest Api, has not been used at my client side implementation and I think that's why I've missed all of your comments here. Probably should have wrote some test cases.

File: synapse/rest/client/v1/admin.py (hunk 3 above)
Comment (morteza-araby) [inline synapse/rest/client/v1/admin.py]: you are right.

File: synapse/rest/client/v1/admin.py (hunk 4 above)
Comment (morteza-araby) [inline synapse/rest/client/v1/admin.py]: should be removed.

File: synapse/rest/client/v1/admin.py (hunk 2 above)
Comment (richvdh) [reply synapse/rest/client/v1/admin.py]: This is now tracked as https://github.com/matrix-org/synapse/issues/2522

File: synapse/rest/client/v1/admin.py (hunk 4 above)
Comment (richvdh) [reply synapse/rest/client/v1/admin.py]: yes it should...

Comment (richvdh) [reply synapse/rest/client/v1/admin.py]: this finally got fixed by #4965

**Pull request merged**